File: challenge-234/steven-wilson/python/ch-01.py
#!/usr/bin/env python3

from collections import Counter


def common_characters(*words):
    """return all alphabetic characters that show up in all words 
    including duplicates
    >>> common_characters("java", "javascript", "julia")
    ('j', 'a')
    >>> common_characters("bella", "label", "roller")
    ('e', 'l', 'l')
    >>> common_characters("cool", "lock", "cook")
    ('c', 'o')
    """
    words = list(words)
    common = list(words.pop(0))
    count_common = Counter(common)

    for word in words:
        count_char = Counter(word)
        common_copy = common.copy()
        for char in common_copy:
            if char not in word:
                common.remove(char)
            elif count_common[char] > count_char[char]:
                count_common[char] -= 1
                common.reverse()
                common.remove(char)
                common.reverse()

    return tuple(common)



if __name__ == "__main__":
    import doctest

    doctest.testmod()

File: challenge-234/steven-wilson/javascript/ch-01.js
"use strict";


function commonCharacters(...words) {
    let common = words.shift();
    let countCommon = new Map();
    for (let character of common) {
        countCommon.set( character, (countCommon.get(character) ?? 0) + 1 );
    }
    for (let word of words) {
        let countCharacters = new Map();
        for (let character of word) {
            countCharacters.set( character, (countCharacters.get(character) ?? 0) + 1);
        }
        for (let character of common) {
            if( word.indexOf(character) == -1 ) {
                let last = common.indexOf(character);
                common = common.slice(0, last) + common.slice(last+1);
            }
            else if (countCommon.get(character) > countCharacters.get(character)){
                countCommon.set( character, (countCommon.get(character) - 1));
                let last = common.lastIndexOf(character);
                common = common.slice(0, last) + common.slice(last+1);
            }
        }
    }
    return common.split("");
}
